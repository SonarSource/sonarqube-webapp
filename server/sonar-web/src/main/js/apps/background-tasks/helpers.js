/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ONE_SECOND = 1000;
const ONE_MINUTE = 60 * ONE_SECOND;

function format (int, suffix) {
  return `${int}${suffix}`;
}

export function formatDuration (value) {
  if (!value) {
    return '';
  }
  if (value >= ONE_MINUTE) {
    const minutes = Math.round(value / ONE_MINUTE);
    return format(minutes, 'min');
  } else if (value >= ONE_SECOND) {
    const seconds = Math.round(value / ONE_SECOND);
    return format(seconds, 's');
  } else {
    return format(value, 'ms');
  }
}
