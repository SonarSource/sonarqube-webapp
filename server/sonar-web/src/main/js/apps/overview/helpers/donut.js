import d3 from 'd3';
import React from 'react';

let Sector = React.createClass({
  render() {
    let arc = d3.svg.arc()
        .outerRadius(this.props.radius)
        .innerRadius(this.props.radius - this.props.thickness);
    return <path d={arc(this.props.data)} style={{ fill: this.props.fill }}/>;
  }
});

export default React.createClass({
  getDefaultProps() {
    return {
      size: 30,
      thickness: 6
    };
  },

  render() {
    let radius = this.props.size / 2;
    let pie = d3.layout.pie()
        .sort(null)
        .value(d => d.value);
    let data = this.props.data;
    let sectors = pie(data).map((d, i) => {
      return <Sector key={i} data={d} fill={data[i].fill} radius={radius} thickness={this.props.thickness}/>;
    });
    return <svg width={this.props.size} height={this.props.size}>
      <g transform={`translate(${radius}, ${radius})`}>{sectors}</g>
    </svg>;
  }
});
