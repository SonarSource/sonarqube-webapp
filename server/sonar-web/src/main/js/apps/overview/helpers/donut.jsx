import React from 'react';

const Sector = React.createClass({
  render() {
    const arc = d3.svg.arc()
        .outerRadius(this.props.radius)
        .innerRadius(this.props.radius - this.props.thickness);
    return <path d={arc(this.props.data)} style={{ fill: this.props.fill }}></path>;
  }
});

export default React.createClass({
  getDefaultProps() {
    return {
      size: 30,
      thickness: 6
    };
  },

  render() {
    const radius = this.props.size / 2;
    const pie = d3.layout.pie().sort(null)
        .value(d => {
          return d.value
        });
    const data = this.props.data;
    const sectors = pie(data).map((d, i) => {
      return <Sector
          key={i}
          data={d}
          fill={data[i].fill}
          radius={radius}
          thickness={this.props.thickness}/>;
    });
    return (
        <svg width={this.props.size} height={this.props.size}>
          <g transform={`translate(${radius}, ${radius})`}>{sectors}</g>
        </svg>
    );
  }
});
