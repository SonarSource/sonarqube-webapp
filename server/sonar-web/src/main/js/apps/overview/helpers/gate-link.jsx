import React from 'react';

export default React.createClass({
  render() {
    const url = `${baseUrl}/quality_gates/show/${this.props.gate}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
