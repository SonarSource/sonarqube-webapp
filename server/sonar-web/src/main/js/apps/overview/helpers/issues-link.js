import React from 'react';

export default React.createClass({
  render() {
    let params = Object.keys(this.props.params).map((key) => {
          return `${key}=${encodeURIComponent(this.props.params[key])}`;
        }).join('|'),
        url = `${baseUrl}/component_issues/index?id=${encodeURIComponent(this.props.component)}#${params}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
