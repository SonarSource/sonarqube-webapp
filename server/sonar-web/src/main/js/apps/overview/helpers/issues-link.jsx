import React from 'react';

export default React.createClass({
  render: function () {
    var params = Object.keys(this.props.params).map((key) => {
      return `${key}=${encodeURIComponent(this.props.params[key])}`;
    }).join('|');
    var url = `${baseUrl}/component_issues/index?id=${encodeURIComponent(this.props.component)}#${params}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
