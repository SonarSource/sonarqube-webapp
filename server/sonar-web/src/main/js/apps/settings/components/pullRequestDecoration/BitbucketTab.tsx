/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import {
  createBitbucketConfiguration,
  updateBitbucketConfiguration
} from '../../../../api/almSettings';
import { BitbucketBindingDefinition } from '../../../../types/alm-settings';
import BitbucketTabRenderer from './BitbucketTabRenderer';

interface Props {
  definitions: BitbucketBindingDefinition[];
  loading: boolean;
  onDelete: (definitionKey: string) => void;
  onUpdateDefinitions: () => void;
}

interface State {
  editedDefinition?: BitbucketBindingDefinition;
  projectCount?: number;
}

export default class BitbucketTab extends React.PureComponent<Props, State> {
  mounted = false;
  state: State = {};

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleCancel = () => {
    this.setState({
      editedDefinition: undefined
    });
  };

  handleCreate = () => {
    this.setState({ editedDefinition: { key: '', url: '', personalAccessToken: '' } });
  };

  handleEdit = (definitionKey: string) => {
    const editedDefinition = this.props.definitions.find(d => d.key === definitionKey);
    this.setState({ editedDefinition });
  };

  handleSubmit = (config: BitbucketBindingDefinition, originalKey: string) => {
    const call = originalKey
      ? updateBitbucketConfiguration({ newKey: config.key, ...config, key: originalKey })
      : createBitbucketConfiguration(config);
    return call
      .then(() => {
        if (this.mounted) {
          this.setState({ editedDefinition: undefined });
        }
      })
      .then(this.props.onUpdateDefinitions);
  };

  render() {
    const { definitions, loading } = this.props;
    const { editedDefinition } = this.state;
    return (
      <BitbucketTabRenderer
        definitions={definitions}
        editedDefinition={editedDefinition}
        loading={loading}
        onCancel={this.handleCancel}
        onCreate={this.handleCreate}
        onDelete={this.props.onDelete}
        onEdit={this.handleEdit}
        onSubmit={this.handleSubmit}
      />
    );
  }
}
