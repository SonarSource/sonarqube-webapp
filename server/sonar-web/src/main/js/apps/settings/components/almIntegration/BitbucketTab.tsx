/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { Link } from 'react-router';
import { translate } from 'sonar-ui-common/helpers/l10n';
import {
  createBitbucketConfiguration,
  updateBitbucketConfiguration
} from '../../../../api/alm-settings';
import {
  AlmKeys,
  AlmSettingsBindingStatus,
  BitbucketBindingDefinition
} from '../../../../types/alm-settings';
import AlmTab from './AlmTab';
import BitbucketForm from './BitbucketForm';

export interface BitbucketTabProps {
  definitions: BitbucketBindingDefinition[];
  definitionStatus: T.Dict<AlmSettingsBindingStatus>;
  loadingAlmDefinitions: boolean;
  loadingProjectCount: boolean;
  multipleAlmEnabled: boolean;
  onCheck: (definitionKey: string) => void;
  onDelete: (definitionKey: string) => void;
  onUpdateDefinitions: () => void;
}

export default function BitbucketTab(props: BitbucketTabProps) {
  const {
    multipleAlmEnabled,
    definitions,
    definitionStatus,
    loadingAlmDefinitions,
    loadingProjectCount
  } = props;

  return (
    <div className="bordered">
      <AlmTab
        alm={AlmKeys.Bitbucket}
        createConfiguration={createBitbucketConfiguration}
        defaultBinding={{ key: '', url: '', personalAccessToken: '' }}
        definitions={definitions}
        definitionStatus={definitionStatus}
        form={childProps => <BitbucketForm {...childProps} />}
        help={
          <>
            <h3>{translate('onboarding.create_project.pat_help.title')}</h3>

            <p className="big-spacer-top">
              {translate('settings.almintegration.bitbucket.help_1')}
            </p>

            <ul className="big-spacer-top list-styled">
              <li>{translate('settings.almintegration.bitbucket.help_2')}</li>
              <li>{translate('settings.almintegration.bitbucket.help_3')}</li>
            </ul>

            <p className="big-spacer-top big-spacer-bottom">
              <Link target="_blank" to="/documentation/analysis/pr-decoration/">
                {translate('learn_more')}
              </Link>
            </p>
          </>
        }
        loadingAlmDefinitions={loadingAlmDefinitions}
        loadingProjectCount={loadingProjectCount}
        multipleAlmEnabled={multipleAlmEnabled}
        onCheck={props.onCheck}
        onDelete={props.onDelete}
        onUpdateDefinitions={props.onUpdateDefinitions}
        updateConfiguration={updateBitbucketConfiguration}
      />
    </div>
  );
}
