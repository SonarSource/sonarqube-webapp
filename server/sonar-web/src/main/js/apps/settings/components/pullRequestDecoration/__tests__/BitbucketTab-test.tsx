/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import {
  createBitbucketConfiguration,
  updateBitbucketConfiguration
} from '../../../../../api/almSettings';
import { mockBitbucketDefinition } from '../../../../../helpers/mocks/alm-settings';
import BitbucketTab from '../BitbucketTab';

jest.mock('../../../../../api/almSettings', () => ({
  countBindedProjects: jest.fn().mockResolvedValue(2),
  createBitbucketConfiguration: jest.fn().mockResolvedValue({}),
  deleteConfiguration: jest.fn().mockResolvedValue({}),
  updateBitbucketConfiguration: jest.fn().mockResolvedValue({})
}));

beforeEach(() => {
  jest.clearAllMocks();
});

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('should handle cancel', async () => {
  const wrapper = shallowRender();

  wrapper.setState({
    editedDefinition: mockBitbucketDefinition()
  });

  wrapper.instance().handleCancel();

  await waitAndUpdate(wrapper);

  expect(wrapper.state().editedDefinition).toBeUndefined();
});

it('should create config', async () => {
  const onUpdateDefinitions = jest.fn();
  const config = mockBitbucketDefinition();
  const wrapper = shallowRender({ onUpdateDefinitions });
  wrapper.setState({ editedDefinition: config });

  await wrapper.instance().handleSubmit(config, '');

  expect(createBitbucketConfiguration).toBeCalledWith(config);
  expect(onUpdateDefinitions).toBeCalled();
  expect(wrapper.state().editedDefinition).toBeUndefined();
});

it('should update config', async () => {
  const onUpdateDefinitions = jest.fn();
  const config = mockBitbucketDefinition();
  const wrapper = shallowRender({ onUpdateDefinitions });
  wrapper.setState({ editedDefinition: config });

  await wrapper.instance().handleSubmit(config, 'originalKey');

  expect(updateBitbucketConfiguration).toBeCalledWith({
    newKey: 'key',
    ...config,
    key: 'originalKey'
  });
  expect(onUpdateDefinitions).toBeCalled();
  expect(wrapper.state().editedDefinition).toBeUndefined();
});

it('should handle create', async () => {
  const wrapper = shallowRender();
  wrapper.instance().handleCreate();
  await waitAndUpdate(wrapper);
  expect(wrapper.state().editedDefinition).toEqual({ key: '', url: '', personalAccessToken: '' });
});

it('should handle edit', async () => {
  const config = {
    key: 'key',
    url: 'url',
    personalAccessToken: 'PAT'
  };
  const wrapper = shallowRender({ definitions: [config] });
  wrapper.instance().handleEdit(config.key);
  await waitAndUpdate(wrapper);
  expect(wrapper.state().editedDefinition).toEqual(config);
});

function shallowRender(props: Partial<BitbucketTab['props']> = {}) {
  return shallow<BitbucketTab>(
    <BitbucketTab
      definitions={[]}
      loading={false}
      onDelete={jest.fn()}
      onUpdateDefinitions={jest.fn()}
      {...props}
    />
  );
}
