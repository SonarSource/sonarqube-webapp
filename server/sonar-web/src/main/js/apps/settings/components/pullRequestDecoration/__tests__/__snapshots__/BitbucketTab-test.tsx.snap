// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<AlmTab
  additionalColumnsHeaders={
    Array [
      "settings.pr_decoration.table.column.bitbucket.url",
    ]
  }
  additionalColumnsKeys={
    Array [
      "url",
    ]
  }
  alm="bitbucket"
  createConfiguration={[Function]}
  defaultBinding={
    Object {
      "key": "",
      "personalAccessToken": "",
      "url": "",
    }
  }
  definitions={
    Array [
      Object {
        "key": "key",
        "personalAccessToken": "asdf1234",
        "url": "http://bbs.enterprise.com",
      },
    ]
  }
  form={[Function]}
  loading={false}
  multipleAlmEnabled={true}
  onDelete={[MockFunction]}
  onUpdateDefinitions={[MockFunction]}
  updateConfiguration={[Function]}
/>
`;
